database.c

```diff
@@ -531,7 +531,7 @@ _db_query_children (GNode *n, struct database_node *parent, GNode *query)
     /* This check doesn't need a * match - they get picked up with the
      * _db_add_children below.
      */
-    if (parent->length &&
+    if (parent->length && !(g_node_first_child(query) && g_node_first_child(query)->data) &&
         (((char*)query->data)[0] == '\0' || strcmp (query->data, parent->hashtree_node.key) == 0))
```

Comment:
This now says: "add path if there is info, it doesn't have children and the search was for the key or the whole directory". It would probably be best to make the new check another if statement so you can lay it out in comments. This stuff was a bit touchy to get right.

if (g_node_first_child(query) && g_node_first_child(query)->data)
{
  debug(skip past this - children are more important)
}
else if (what was there)

Reply:
updated